plugins/candela/web_client/views/CandelaWidget.js

```diff
@@ -44,61 +45,77 @@ var CandelaWidget = View.extend({
-                    };
-                    events.trigger('g:alert', info);
-                    console.error(error);
-                    return;
-                }
+        let cleanup = function () {
+            this.$('.g-item-candela')
+                .remove();
+        };
+
+        if (this.accessLevel < AccessType.READ) {
+            cleanup();
+            return;
+        }
+
+        let parser = null;
+        let name = this.item.get('name').toLowerCase();
+        if (name.endsWith('.csv')) {
+            parser = datalib.csv;
+        } else if (name.endsWith('.tsv') || name.endsWith('.tab')) {
+            parser = datalib.tsv;
+        } else {
+            cleanup();
+            return;
+        }
+
+        this.$el.html(CandelaWidgetTemplate({
+            components: this._components
+        }));
+        this.parametersView.setElement($('.g-item-candela-parameters'));
+        parser(this.item.downloadUrl(), (error, data) => {
+            if (error) {
+                let info = {
+                    text: 'An error occurred while attempting to read and ' +
+                          'parse the data file. Details have been logged in the console.',
+                    type: 'danger',
+                    timeout: 5000,
+                    icon: 'attention'
+                };
+                events.trigger('g:alert', info);
+                console.error(error);
+                return;
+            }
 
-                datalib.read(data, {parse: 'auto'});
-
-                // Vega has issues with empty-string fields and fields with dots, so rename those.
-                let rename = [];
-                for (let key in data.__types__) {
-                    if (data.__types__.hasOwnProperty(key)) {
-                        if (key === '') {
-                            rename.push(['', 'id']);
-                        } else if (key.indexOf('.') >= 0) {
-                            rename.push([key, key.replace(/\./g, '_')]);
-                        }
+            datalib.read(data, {parse: 'auto'});
+
+            // Vega has issues with empty-string fields and fields with dots, so rename those.
+            let rename = [];
+            for (let key in data.__types__) {
+                if (data.__types__.hasOwnProperty(key)) {
+                    if (key === '') {
+                        rename.push(['', 'id']);
+                    } else if (key.indexOf('.') >= 0) {
+                        rename.push([key, key.replace(/\./g, '_')]);
                     }
                 }
+            }
 
-                rename.forEach((d) => {
-                    data.__types__[d[1]] = data.__types__[d[0]];
-                    delete data.__types__[d[0]];
-                    data.forEach((row) => {
-                        row[d[1]] = row[d[0]];
-                        delete row[d[0]];
-                    });
+            rename.forEach((d) => {
+                data.__types__[d[1]] = data.__types__[d[0]];
```

Comment:
The `d[0]` issue should be resolved.

`__types__` is added to the parsed array by the `datalib` parser methods, and includes inferred type information for each field. For some reason the library does this instead of returning another value.